File: Trees/BinaryTrees/CheckIfTreeIsBalanced.cpp
#include <bits/stdc++.h>
using namespace std;

typedef struct TreeNode{
    int val;
    TreeNode* left,*right;

    TreeNode(){
    }

    TreeNode(int val){
        this->val = val;
        left = right = NULL;
    }
}TreeNode;

class Util{
	public:
    static TreeNode* getTree(){
        TreeNode* root = new TreeNode(10);
        root->left = new TreeNode(20);
        root->right = new TreeNode(30);

        root->left->left = new TreeNode(40);
        root->left->right = new TreeNode(50);
        root->left->right->left = new TreeNode(60);
        root->left->right->right = new TreeNode(70);
        return root;
    }

    static int height(TreeNode* root){
        if(root == NULL)    return -1;

        int lh = height(root->left);
        int rh = height(root->right);

        return max(lh,rh) + 1;
    }

    static bool isBalanced(TreeNode* root){
        if(root == nullptr) return true;

        bool isLeft = isBalanced(root->left);
        bool isRight = isBalanced(root->right);
        
        if(!isLeft || !isRight) return false;

        int lh = height(root->left);
        int rh = height(root->right);

        bool isReallyBalanced = (abs(lh - rh) < 2);

        return isLeft && isRight && isReallyBalanced;
    }

    static pair<bool,int> isBalancedOptimal(TreeNode* root){
        if(root == nullptr) return {true,-1};

        pair<bool,int> leftHeight = isBalancedOptimal(root->left);
        pair<bool,int> rightHeight = isBalancedOptimal(root->right);
        
        if(!leftHeight.first || !rightHeight.first) return {false,-1};

        int currentNodeHeight = max(rightHeight.second,leftHeight.second) + 1;
        bool isCurrentNodeBalanced = abs(leftHeight.second - rightHeight.second) < 2;

        return {isCurrentNodeBalanced,currentNodeHeight};
    }

};

int main(){
    TreeNode* root = Util::getTree();
    cout<<Util::isBalancedOptimal(root);
    return 0;
}

File: Trees/BinaryTrees/CheckIfTreeIsBalanced.java
package Trees.BinaryTrees;

public class CheckIfTreeIsBalanced {

    private static boolean isBalancedSolutionOne(TreeNode root) {
        if (root == null) return true;

        boolean isLeftBalanced = isBalancedSolutionOne(root.left);
        boolean isRightBalanced = isBalancedSolutionOne(root.right);

        if (!isLeftBalanced || !isRightBalanced) return false;

        // now check if current node is balanced ?
        int leftHeight = TreeOperations.getTreeHeight(root.left);
        int rightHeight = TreeOperations.getTreeHeight(root.right);

        int finalheight = Math.abs(leftHeight - rightHeight);
        return finalheight <= 1;
    }

    private static class Ans {
        int height;
        boolean isBalanced;

        public Ans(int height, boolean isBalanced) {
            this.height = height;
            this.isBalanced = isBalanced;
        }

        @Override
        public String toString() {
            return "Ans{" +
                    "height=" + height +
                    ", isBalanced=" + isBalanced +
                    '}';
        }
    }

    private static Ans isBalancedSolutionTwo(TreeNode root) {
        if (root == null) return new Ans(0, true);

        Ans leftSubtree = isBalancedSolutionTwo(root.left);
        Ans rightSubtree = isBalancedSolutionTwo(root.right);

        // if any of subtree is not balanced then there is no point in figuring out whether current or above node are balanced or not, hence returned false.
        if (!leftSubtree.isBalanced || !rightSubtree.isBalanced) return new Ans(-1, false);

        int finalHeight = Math.abs(leftSubtree.height - rightSubtree.height);
        boolean isCurrentNodeBalanced = (finalHeight <= 1 && leftSubtree.isBalanced && rightSubtree.isBalanced);
        int currentNodeHeight = Math.max(leftSubtree.height, rightSubtree.height) + 1;

        return new Ans(currentNodeHeight, isCurrentNodeBalanced);
    }


    private static Ans isBalancedBST(TreeNode root) {
        if (root == null) return new Ans(-1, true);

        Ans leftAns = isBalancedBST(root.left);
        Ans rightAns = isBalancedBST(root.right);

        if (!leftAns.isBalanced || !rightAns.isBalanced) return new Ans(-1, false);

        int leftHeight = leftAns.height;
        int rightHeight = rightAns.height;

        int finalHeight = Math.abs(leftHeight - rightHeight);
        boolean isBalanced = (finalHeight <= 1);

        return new Ans(Math.max(leftHeight, rightHeight) + 1, isBalanced);
    }

    public static void main(String[] args) {
        TreeNode root = TreeNode.getInstance();
        System.out.println("isBalanced ? : " + isBalancedSolutionTwo(root));
    }
}
